External prog for html stripping

eldipa opened the pull request:
Proposed implementation for #5

eldipa committed:
Explicit call to HTML parser's close and reset

A call to feed() does not guaranties that all the data will be processed
by the parser: some data may be left unless close() is called.

And because the same parser is reused, a call to reset() will ensure a
clean state for each post stripping.

eldipa committed:
Call an external program for HTML stripping

If -S is given, the argument will be a program that will read HTML from
stdin and it will output a stripped version of it.

The program is executed by the shell so we allow the user to execute a
piped command for example.

-S implies -s and if both are given, -S will have priority over -s.

The input/output is stored as bytes and converter to UTF-8 by us.

eldipa committed:
Add a test for ExternalHTMLStripper

sulami commented:
I like the idea of this.

I'll have a close look tomorrow, but I can already see that this will also require a quick README update.

eldipa commented:
Cool!, send me your feed back and indeed, I totally forgot the README :D Perhaps adding the `-S` option next with the rest of the options and a small example in a separated section?

sulami commented on scripts/feed2maildir:
```diff
@@ -40,7 +43,11 @@ def main():
     else:
         maildir = config['maildir']
 
+    if args['strip_program']:
+        args['s'] = True
```
Rather than modifying the `args`, I'd rather just introduce a `strip` variable and set that to `args['s'] or args['strip_program']`.

sulami commented on feed2maildir/converter.py:
```diff
@@ -55,6 +57,28 @@ def get_data(self):
+        self.raw_data.append(data)
+
+    def close(self):
+        p = Popen(self.strip_program, stdin=PIPE, stdout=PIPE, shell=True)
```
Any particular reason for using `Popen` over `subprocess`?

sulami replied:
Also, I think we might want to handle common failure cases, like program not found and non-zero exit codes.

eldipa replied:
`subprocess.Popen` is flexible enough to allow me to send it data and receive from it the output via the `communicate` method.

If the program does not exist, python will race an exception, however if it fails with a non-zero exit code the user will not notice it. I'm going to add a check for that, print the stderr and raise an exception. Good catch!

eldipa replied:
I did an extra check and if a program does not exist, `Popen` will not fail. This is because the program executed is actually a shell (`shell=True`). Checking for the exit code later will catch that error case.

sulami reviewed (commented):
I finally got around to adding CI to this, so you'll have to rebase onto master.

eldipa committed:
Use a 'strip' variable instead of 'args'

eldipa committed:
Check for non-zero return code

After the call, check the exit code and it is non-zero, raise a
CalledProcessError like the subprocess.run function does.

On failure print the standard error of the failed program. When we drop
support of 2.7 and 3.2 and lower, we can add the standard error as part
of the CalledProcessError and handle it later.

eldipa committed:
Document the -S option

eldipa commented on README.rst:
```diff
@@ -51,8 +51,27 @@ There are a bunch of command-line arguments to overwrite the config file:
+::
+
+    feed2maildir -S 'pandoc --from html --to markdown_strict'
+
```
I'm not an English native speaker, so these would need a second look for misspelling or wrong expressions.

sulami commented:
I know I'm rather late on this, but this looks great, thank you!
